Multiple coupon support in Order creation

samiuelson opened the pull request:
## Multiple coupon support in Order creation
<!-- Remember about a good descriptive title. -->

Closes: #9276 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
The flow of adding a coupon to an order was updated to allow adding multiple coupons. See the video below.

1. Added new applied coupon list screen redirecting to the coupon details/edition screen.
In case there are no coupons in the order we redirect to coupon details directly to remove the redundant (coupon list) step.
2. Modified coupon addition and deletion logic in `OrderCreateEditViewModel` to support multiple coupon lines.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
You'll need to create at least two coupons in wp-admin.

1. Create new order and try to add first coupon.
2. Verify you are redirected directly to coupon entry screen and that coupon is applied correctly to the order.
3. Try adding the second coupon by clicking coupons label and observe we show list of already applied coupons first. Click on the "+" FAB.
4. Enter the second coupon and verify both coupons are applied correctly to the order.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
https://github.com/woocommerce/woocommerce-android/assets/4527432/af7fe767-0992-486d-9cff-dd4cd0b890ea

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Implement multiple coupon support

1. Added new applied coupons list screen redirecting to coupon details screen
2. Modified coupon addition and deletion logic in `OrderCreateEditViewModel` to support multiple coupon lines
3. Improved existing coupon removal logic

samiuelson committed:
Fix detekt issue

samiuelson committed:
Update unit tests

samiuelson committed:
Merge branch 'trunk' into 9276-multiple-coupons-in-order

samiuelson committed:
Update RELEASE-NOTES

samiuelson committed:
Improve naming style

samiuelson committed:
Simplify code

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/coupon/OrderCouponListScreen.kt:
```diff
@@ -0,0 +1,138 @@
+                style = MaterialTheme.typography.subtitle1,
+                color = MaterialTheme.colors.onSurface,
+            )
+            Divider(
```
Divider has only start offset, is this intentional? Also, we probably don't want to show that after the last item?

<img width="376" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/ae27869b-633a-4634-b6fa-9dc77aa434d6">

samiuelson replied:
> Divider has only start offset, is this intentional?

I agree this is a bit weird UI at first sight, but that's how dividers look like in other parts of the app (e.g. product list, coupon list) and that's also recommended by Google https://m2.material.io/components/dividers

> Also, we probably don't want to show that after the last item?

Hiding the last divider in fd58549.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/coupon/OrderCouponListScreen.kt:
```diff
@@ -0,0 +1,138 @@
+package com.woocommerce.android.ui.orders.creation.coupon
```
np: maybe better to put "list" and "edition" into different packages inside of "coupon"?

Also, should it be `editing`, instead of `edition`?

samiuelson replied:
1. Renamed `edition` to `edit` in a0029ac - I think this will fit even better here.
2. Organized into separate packages in e7a5486

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/coupon/OrderCreateCouponListViewModel.kt:
```diff
@@ -0,0 +1,37 @@
+    ).asLiveData()
+
+    fun onAddCouponClicked() {
+        triggerEvent(OrderCreateEditNavigationTarget.EditCoupon(null))
```
np: `EditCoupon` that takes `null` code semantically feels wrong. What if we will have `object AddCoupon` and `EditCoupon(code:String)`

Also, maybe use own events here instead of the `OrderCreateEditNavigationTarget` to keep a bit more separation between screens

samiuelson replied:
Makes sense! Fixed with b4b14d3

kidinov commented:
Shell, do we show an error message to a user when an incorrect coupon is entered? Right now, it's not clear if the code is incorrect, the network request failed, or something else went wrong.

Also, shell it be single-line test input?

<img width="385" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/a996156d-a225-475a-b432-c576a8bc3abd">

kidinov commented:
There is a crash:

```
java.lang.reflect.InvocationTargetException
at java.lang.reflect.Method.invoke(Native Method)
at com.woocommerce.android.viewmodel.NavArgsLazy.getValue(SavedStateHandleExt.kt:43)
at com.woocommerce.android.viewmodel.NavArgsLazy.getValue(SavedStateHandleExt.kt:24)
at com.woocommerce.android.ui.orders.creation.OrderCreateEditViewModel.getArgs(OrderCreateEditViewModel.kt:142)
at com.woocommerce.android.ui.orders.creation.OrderCreateEditViewModel.<init>(OrderCreateEditViewModel.kt:143)
at com.woocommerce.android.DaggerWooCommerceDebug_HiltComponents_SingletonC$ViewModelCImpl$SwitchingProvider.get0(DaggerWooCommerceDebug_HiltComponents_SingletonC.java:6179)
at com.woocommerce.android.DaggerWooCommerceDebug_HiltComponents_SingletonC$ViewModelCImpl$SwitchingProvider.get(DaggerWooCommerceDebug_HiltComponents_SingletonC.java:6432)
at dagger.hilt.android.internal.lifecycle.HiltViewModelFactory$1.create(HiltViewModelFactory.java:102)
at androidx.lifecycle.AbstractSavedStateViewModelFactory.create(AbstractSavedStateViewModelFactory.java:89)
at dagger.hilt.android.internal.lifecycle.HiltViewModelFactory.create(HiltViewModelFactory.java:114)
at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.kt:187)
at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.kt:153)
at androidx.lifecycle.ViewModelLazy.getValue(ViewModelLazy.kt:53)
at androidx.lifecycle.ViewModelLazy.getValue(ViewModelLazy.kt:35)
at com.woocommerce.android.ui.orders.creation.coupon.OrderCreateCouponEditionFragment.getSharedViewModel(OrderCreateCouponEditionFragment.kt:51)
at com.woocommerce.android.ui.orders.creation.coupon.OrderCreateCouponEditionFragment.access$getSharedViewModel(OrderCreateCouponEditionFragment.kt:26)
at com.woocommerce.android.ui.orders.creation.coupon.OrderCreateCouponEditionFragment$onViewCreated$1.invoke(OrderCreateCouponEditionFragment.kt:79)
at com.woocommerce.android.ui.orders.creation.coupon.OrderCreateCouponEditionFragment$onViewCreated$1.invoke(OrderCreateCouponEditionFragment.kt:76)
at com.woocommerce.android.ui.orders.creation.coupon.OrderCreateCouponEditionFragment.onViewCreated$lambda$1(OrderCreateCouponEditionFragment.kt:76)
at com.woocommerce.android.ui.orders.creation.coupon.OrderCreateCouponEditionFragment.$r8$lambda$ktW5WM4XYJuXO2lkP29FAfDgL_A(Unknown Source:0)
at com.woocommerce.android.ui.orders.creation.coupon.OrderCreateCouponEditionFragment$$ExternalSyntheticLambda1.onChanged(Unknown Source:2)
at com.woocommerce.android.viewmodel.MultiLiveEvent$observe$1.invoke(MultiLiveEvent.kt:39)
at com.woocommerce.android.viewmodel.MultiLiveEvent$observe$1.invoke(MultiLiveEvent.kt:36)
at com.woocommerce.android.viewmodel.MultiLiveEvent.observe$lambda$0(MultiLiveEvent.kt:36)
at com.woocommerce.android.viewmodel.MultiLiveEvent.$r8$lambda$EMtCJbslvEzla8wHlyfjSLoOiQE(Unknown Source:0)
at com.woocommerce.android.viewmodel.MultiLiveEvent$$ExternalSyntheticLambda0.onChanged(Unknown Source:2)
at androidx.lifecycle.LiveData.considerNotify(LiveData.java:133)
at androidx.lifecycle.LiveData.dispatchingValue(LiveData.java:151)
at androidx.lifecycle.LiveData.setValue(LiveData.java:309)
at androidx.lifecycle.MutableLiveData.setValue(MutableLiveData.java:50)
at com.woocommerce.android.viewmodel.MultiLiveEvent.setValue(MultiLiveEvent.kt:52)
at com.woocommerce.android.viewmodel.MultiLiveEvent.setValue(MultiLiveEvent.kt:30)
at com.woocommerce.android.viewmodel.ScopedViewModel.triggerEvent(ScopedViewModel.kt:37)
at com.woocommerce.android.ui.orders.creation.coupon.OrderCreateCouponEditionViewModel.onDoneClicked(OrderCreateCouponEditionViewModel.kt:56)
at com.woocommerce.android.ui.orders.creation.coupon.OrderCreateCouponEditionViewModel$onDoneClicked$1.invokeSuspend(Unknown Source:14)
at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
```

Check the video for reproduction steps

https://github.com/woocommerce/woocommerce-android/assets/4923871/1ab674ba-ebcf-48e1-9c74-e086396baae5

I feel that shared `ViewModel` is a source of many troubles for us lately. We may want to avoid it and pass data via bundles from screen to screen

kidinov commented:
When you edit a coupon, I assume it will edit it, not add a new one


https://github.com/woocommerce/woocommerce-android/assets/4923871/b3b5017d-0e89-4ee2-839d-cfc1982c54a9

kidinov commented:
As a general note - this maybe not be clear to a user that they can add another coupon by clicking on the one that has already been added. Maybe keep showing the "add coupon" button even if 1 or more coupons are added?

<img width="380" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/9258100c-6d8a-412a-955a-c9129c4397aa">

kidinov reviewed (commented):
Thank you for working on that!

I left a couple of comments here; please take a look.

Also, please confirm this temporary solution, and eventually, we will show a list of coupons with info about them from which users can select, right?

samiuelson committed:
Hide divider after the last coupon

samiuelson committed:
Rename `OrderCreateCouponEdition` classes to `OrderCreateCouponEdit`

samiuelson committed:
Organize order's coupon files in separate packages

samiuelson committed:
Clean up navigation events model

samiuelson committed:
Merge remote-tracking branch 'origin/trunk' into 9276-multiple-coupons-in-order

samiuelson committed:
Stop using shared view model

1. Pass coupon add/edit/remove result via bundle
2. Distinguish between coupon editing and adding - don't add new coupon if it was edited

samiuelson committed:
Make coupon text field single line

samiuelson commented:
> caf8dec

samiuelson committed:
Improve coupon update conditions

samiuelson committed:
Make +Add coupon button always available

samiuelson commented:
Thanks for the great review, @kidinov! I've addressed your comments.

> Maybe keep showing the "add coupon" button even if 1 or more coupons are added?

Done 7b62d5e

> When you edit a coupon, I assume it will edit it, not add a new one

Fixed with 46b28ec

> There is a crash. The shared ViewModel is a source of many troubles for us lately. 

Thanks for raising this. I agree it's an antipattern and is introducing problems. I stopped using the shared view model in the coupon-related classes and implemented passing results as fragment args back and forth instead.

Additionally, I've made the coupon text field a single liner and added an issue to improve validation/network error information showing: https://github.com/woocommerce/woocommerce-android/issues/9314

samiuelson commented:
> Also, please confirm this temporary solution, and eventually, we will show a list of coupons with info about them from which users can select, right?

@kidinov Yes, that is the plan to implement a coupon selector component for easier coupon adding as one of the next improvements.

samiuelson committed:
Fix detekt issues

samiuelson committed:
Clean up redundant code

samiuelson commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/creation/coupon/OrderCreateCouponEditionViewModelTest.kt:
Moved to a new package and file name changed

samiuelson committed:
Merge remote-tracking branch 'origin/trunk' into 9276-multiple-coupons-in-order

samiuelson committed:
Fix detekt issues

kidinov commented:
@samiuelson  thank you for the changes. Looks good now, but there is a small bug I noticed (with don't keep activities)

Notice that the coupons are lost at the end

https://github.com/woocommerce/woocommerce-android/assets/4923871/cbb0959a-7271-4c9c-b4f2-6721ee4093ab

samiuelson committed:
Fix coupon edit/add result processing

samiuelson committed:
Merge remote-tracking branch 'origin/trunk' into 9276-multiple-coupons-in-order

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt
#	WooCommerce/src/main/res/navigation/nav_graph_order_creations.xml

samiuelson committed:
Clean up code

samiuelson commented:
> Looks good now, but there is a small bug I noticed (with don't keep activities). Notice that the coupons are lost at the end.

Good catch @kidinov. Fixed with 800b06b.

samiuelson committed:
Update tests

kidinov commented:
@samiuelson are you sure that this is fixed? Notice that products and coupons are gone


https://github.com/woocommerce/woocommerce-android/assets/4923871/51c8f8d3-1c2d-450f-8bf9-4b4abd76e7ef

kidinov reviewed (approved):
The bug that is still here is not blocking, so I am approving the PR. But if you can, it'll be better to fix it before the cut-off

Looking forward to coupon selection from a list in the next iteration!
